lightning-terminal: fixed building frontend and embedding it into backend

starius opened the pull request:
<!--
^ Please summarise the changes you have done and explain why they are necessary here ^

For package updates please link to a changelog or describe changes, this helps your fellow maintainers discover breaking updates.
For new packages please briefly describe the package or provide a link to its homepage.
-->

## Things done

I used https://github.com/NixOS/nixpkgs/pull/388118 and then fixed building frontend (yarn app in directory "app") and embedding it into backend (`litd` binary) in package `lightning-terminal`.

Previously, the yarn app was not built correctly and the resulting files were not placed where the backend (via Go's `//go:embed`) expects them. As a result, litd served an empty web UI page with a single link to `.gitkeep`, the only file present in `app/build` from Git.
    
Upstream uses a `Dockerfile` and `app/gen_app_docker.sh` to install yarn and populate that directory. This commit replicates that behavior using Nix.
    
Since the app uses yarn v1, `mkYarnPackage` was inappropriate and was replaced with a `stdenv.mkDerivation` using `yarnConfigHook`, `yarnInstallHook`, and `yarnBuildHook`. Linters were disabled to prevent build failures due to findings such as single vs double quotes.
    
The `package.json` file included a `postbuild` step using `git restore`, which failed in the Nix sandbox due to the absence of Git. This step was disabled, as the file it touches is not modified or necessary for the build.
    
To ensure the backend can embed the frontend properly, this commit copies the built frontend files into app/build inside the backend's source tree during `postUnpack`. An assertion was added to check that `index.html` exists there to prevent this issue from silently recurring.

CC @HannahMR

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#linking-nixos-module-tests-to-a-package) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [Nixpkgs 25.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/doc/release-notes/rl-2511.section.md) (or backporting [24.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/release-notes/rl-2411.section.md) and [25.05](https://github.com/NixOS/nixpkgs/blob/master/doc/manual/release-notes/rl-2505.section.md) Nixpkgs Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
- [NixOS 25.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2511.section.md) (or backporting [24.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2411.section.md) and [25.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2505.section.md) NixOS Release notes)
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md), [pkgs/README.md](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md), [maintainers/README.md](https://github.com/NixOS/nixpkgs/blob/master/maintainers/README.md) and other contributing documentation in corresponding paths.

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!

List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#reviewing-contributions
-->

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

ethancedwards8 committed:
lightning-terminal: remove package.json

package.json is included upstream and there is no reason for it to be
inlcuded in the nixpkgs repo.

Signed-off-by: Ethan Carter Edwards <ethan@ethancedwards.com>

starius committed:
lightning-terminal: fix frontend build

Previously, the yarn app was not built correctly and the resulting files were
not placed where the backend (via Go's //go:embed) expects them. As a result,
litd served an empty web UI page with a single link to .gitkeep, the only file
present in app/build from Git.

Upstream uses a Dockerfile and app/gen_app_docker.sh to install yarn and
populate that directory. This commit replicates that behavior using Nix.

Since the app uses yarn v1, mkYarnPackage was inappropriate and was replaced
with a stdenv.mkDerivation using yarnConfigHook, yarnInstallHook, yarnBuildHook.
Linters were disabled to prevent build failures due to findings such as single
vs double quotes.

The package.json file included a postbuild step using 'git restore', which
failed in the Nix sandbox due to the absence of Git. This step was disabled,
as the file it touches is not modified or necessary for the build.

To ensure the backend can embed the frontend properly, this commit copies the
built frontend files into app/build inside the backend's source tree during
postUnpack. An assertion was added to check that index.html exists there to
prevent this issue from silently recurring. Also a nix test was added that runs
litd and checks that it serves index.html using curl.

ju1m reviewed (approved):
@starius, great work fixing the integration of `lightning-app`, thank you!
I suggested a few improvements should you want to take the time to integrate them.
Otherwise please tell me and I'll propose them in a followup PR.

starius commented on pkgs/by-name/li/lightning-terminal/package.nix:
```diff
@@ -75,24 +92,46 @@ buildGoModule rec {
+      yarn build
+
+      runHook postBuild
+    '';
```
Applied

starius commented on pkgs/by-name/li/lightning-terminal/package.nix:
```diff
@@ -30,6 +35,18 @@ buildGoModule rec {
+  postUnpack = ''
+    echo "Copying app build output into app/build dir to embed into litd."
+    cp -r ${lightning-app}/* source/app/build/
+
+    echo "Asserting that app/build/index.html exists."
+    if [ ! -f source/app/build/index.html ]; then
+      echo "ERROR: app/build/index.html not found!"
+      exit 1
+    fi
```
I decided to keep this check. It is fast and helps failing fast during the build, not at test phase. I added your HTTP test as well, thank you!

ju1m replied:
As you wish, I just wanted to be sure you knew about `passthru.tests`,  AFAIK they have to be run before merging and some are run by Hydra.

Note that should a NixOS' `services.ligthning-terminal` be written one day, `passthru.tests.litd-app` should be replaced by a `passthru.tests.nixos = nixosTests.ligthning-terminal` using `wait_for_open_port(8080)` instead of `sleep 2`.

starius replied:
Thank you! Now I know about it :)
I'll keep a note for myself in case `services.ligthning-terminal` is written.

starius commented:
@ju1m Thank you for the feedback!
I incorporated your proposals in a separate commit. Please take a look! I'll squash it into the main commit after that. This is in a separate commit just to make it clear what changed.

I also added a version test. Both tests pass locally.

ju1m commented on pkgs/by-name/li/lightning-terminal/package.nix:
```diff
@@ -75,24 +94,65 @@ buildGoModule rec {
+    '';
+  };
+
+  passthru.tests.version = testers.testVersion {
+    package = lightning-terminal;
+    command = "litd --version";
+    version = "litd version ${version} commit=v${version}";
+  };
+
```
There's a shortcut: [`versionCheckHook`](https://nixos.org/manual/nixpkgs/unstable/#versioncheckhook) (see a little bit above in the code).
```suggestion
```

ju1m commented on pkgs/by-name/li/lightning-terminal/package.nix:
```diff
@@ -75,24 +92,46 @@ buildGoModule rec {
     versionCheckHook
   ];
 
```
So, my test proposal is subtly brittle: `litd` fails to bind port 8080/tcp if a program of the build machine already listens on that port. That's because under the hood `testers.runCommand` is a fixed-output derivation, meaning the script is not run in a private network namespace, enabling it to access the Internet for instance.

The usual `runCommandWith` can be used instead, as it's run in a sandbox as expected:
```nix
  passthru.tests.litd-app =
    runCommandWith
      {
        name = "test-litd-app";
        derivationArgs = {
          nativeBuildInputs = [
            curl
            lightning-terminal
          ];
        };
      }
      ''
        litd \
          --uipassword=12345678 \
          --insecure-httplisten=127.0.0.1:8080 \
          --httpslisten= &
        sleep 2
        GETindexHTTPCode=$(curl -o /dev/null -w "%{http_code}" -Lv http://127.0.0.1:8080/index.html)
        if [ "$GETindexHTTPCode" = 200 ]; then
          touch $out
        fi
      '';
```
Again, this is mostly to test that `litd` is not obviously failing to embed `lightning-app`, but if there ever is a `services.litd`, this should be moved to a `nixosTests`.

ju1m commented on pkgs/by-name/li/lightning-terminal/package.nix:
```diff
@@ -75,24 +94,59 @@ buildGoModule rec {
+      cp -r build/* $out/
+    '';
   };
-  outputs = [
```
`outputsToInstall` (below) can be removed along this

starius commented on pkgs/by-name/li/lightning-terminal/package.nix:
```diff
@@ -101,6 +155,5 @@ buildGoModule rec {
     changelog = "https://github.com/lightninglabs/lightning-terminal/releases/tag/v${version}";
     maintainers = with lib.maintainers; [ HannahMR ];
     mainProgram = "litcli";
```
By the way, why the main program is `litcli`? The main daemon is `litd` and `litcli` is a CLI tool to manage it.
